File: jserver-security/src/main/java/com/github/gr3gdev/jserver/security/TokenClientPlugin.java
package com.github.gr3gdev.jserver.security;

import com.github.gr3gdev.jserver.plugin.ServerPlugin;
import com.github.gr3gdev.jserver.security.user.JwtData;
import com.github.gr3gdev.jserver.security.user.UserData;
import io.fusionauth.jwt.Verifier;
import io.fusionauth.jwt.domain.JWT;
import io.fusionauth.jwt.rsa.RSAVerifier;

import java.security.PublicKey;
import java.util.Optional;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * TokenClientPlugin.
 *
 * @author Gregory Tardivel
 */
public class TokenClientPlugin implements ServerPlugin {

    private static final Logger LOGGER = Logger.getLogger(TokenClientPlugin.class.getSimpleName());

    private final PublicKey key;

    public TokenClientPlugin(final PublicKey key) {
        this.key = key;
    }

    public <T extends UserData> Optional<JwtData<T>> getUserData(final String token, final Class<T> clazz) {
        if (token != null) {
            try {
                final Verifier verifier = RSAVerifier.newVerifier(this.key);
                final JWT jwt = JWT.getDecoder().decode(token, verifier);
                return Optional.of(new JwtData<>(jwt, clazz));
            } catch (final Exception exc) {
                LOGGER.log(Level.SEVERE, "JWT UserData error", exc);
            }
        }
        return Optional.empty();
    }
}

File: jserver-security/src/main/kotlin/com/github/gr3gdev/jserver/security/TokenClientPlugin.kt
package com.github.gr3gdev.jserver.security

import com.github.gr3gdev.jserver.logger.Logger
import com.github.gr3gdev.jserver.plugin.ServerPlugin
import com.github.gr3gdev.jserver.security.user.JwtData
import com.github.gr3gdev.jserver.security.user.UserData
import io.jsonwebtoken.Jwts
import java.security.PublicKey
import java.util.*

class TokenClientPlugin(private val key: PublicKey) : ServerPlugin {

    /**
     * Get user data from JWT token.
     */
    fun <T : UserData> getUserData(token: String?, clazz: Class<T>): Optional<JwtData<T>> {
        if (token != null) {
            try {
                val claims = Jwts.parser().setSigningKey(key).parseClaimsJws(token)
                return Optional.of(JwtData(claims.body, clazz))
            } catch (exc: Exception) {
                Logger.error("JWT UserData error", exc)
            }
        }
        return Optional.empty()
    }

}